### ISSUE TITLE
FairOrderedMemoryAwareThreadPoolExecutor has severe memory leak under heavy load

### ISSUE THREAD

**Issue by novos40** [netty/netty#4311]
Netty version 3.10.4 Final

As per your own documentation the order of events on the pipeline is not guaranteed which is definitely the case under heavy load. Therefore when WRITTEN_AMOUNT event comes after CHANNEL_CLOSED event, executor re-registers already closed channel in it's internal map **again** thus producing a memory leak as these channels will not be ever removed from the map and therefore never garbage collected. Using weak keys in the map helps a bit, but not much as cleanup is deferred and may not happen before OOME occurs. Also if the load comes in bursts of high intensity with dead pauses in between, stale reference queue will not be cleaned until next access. By that time it may be already too late as next burst will hijack the CPU bandwidth required for clean up. Reliance on GC to initiate further clean up is a bad idea.
The issue is easily reproduced with apache bench with high concurrency level (e.g. -c 1000 or so).
Reimplementing the doExecute() method the following way

```
    @Override
    protected void
        doExecute(Runnable task)
    {
        if(task instanceof ChannelEventRunnable) 
        {
            ChannelEventRunnable run = (ChannelEventRunnable)task;
            SffEventTask prev, next = new SffEventTask(run);
            Channel channel = run.getEvent().getChannel();
            if(channel.getCloseFuture().isDone())                                   // remove the entry when the channel closes.
                prev = map.remove(channel.getId());
            else
                prev = map.put(channel.getId(), next);
            if(prev != null 
            && uNext.compareAndSet(prev, null, next)) 
                return;
            task = next;
        } 
        doUnorderedExecute(task);
    }
```

solves the problem (and does not require weak keys in the map).

On an unrelated note:
Using private- and package- scoped variables in open source code is an insult to your fellow developers who use and extend your code. 
Instead of simply sub-classing your class and overriding a single method, I had to re-implement the whole class with all it's internals because you use private and package access for no reason. I will also have to do it again if you do any changes and watch manually for any incompatibilities instead of delegating it to the compiler. Please don't use private/package access. Do use protected instead.

Thanks, Oleg

**Comment by normanmaurer** [comment]
@novos40 maybe you want to submit a PR that we can review and just merge to fix it ?

About package private and private scoped things... This is mainly to limit the "public apis" as otherwise it may become very hard to do changes later on without break the API.

**Comment by novos40** [comment]
@normanmaurer

I'm not sure if PR will do you any good as my implementation is rather different from yours in almost every aspect (i.e. coding style, name conventions, variable names, etc.). It would not merge easily if at all. 
On the other hand, the idea of the fix is very simple: instead of pipeline events (which are unreliable in terms of their order), look for closeFuture to be done (which can happen only once) and essentially switch back to unfair mode as after the closure nobody would care about fairness any more. The snippet I've posted does just that.

Netty is a wonderful, very well designed library, so please don't get me wrong about my rant before. However, with all due respect, you would probably never think of (and much less design it for) ways how it can be used and how I use it in my project. For example, there is no leakage whatsoever of any netty's artifacts (e.g. import packages, classes etc) into other components. It is completely componetized and sealed off from the rest of the system (OSGi helps a lot with that). I've also extended it's capabilities quite a bit to accommodate fully dynamic nature of the project. All of this was possible mainly due to ability to extend your code (which I'm very grateful for). There are just a very few very annoying places like this when it's really impossible to do something in a simple way. I, personally, never use private/package access in my code unless I have a specific reason to do so, which I always document in the code. It's always easier simply to seal something off and never think that somebody would use you code in some way other that you made it for. I do believe that using privates is a sure sign of a sloppy and thoughtless programming and I know that you guys are better than that.
As per the protection of public APIs then it's a wrong tool for the job. APIs must be published and documented, everything else is implementation details which are subject to change and it must be made clear, that's it. In my case I don't really care about your APIs at all as it's all encapsulated in my own abstractions. In fact I, don't even use the concept of API as a software integration tool to begin with. I do care, however, about extensibility of your code to suit my needs and requirements (which are not API based)

Thanks again for a very useful library!
--Oleg

**Comment by atellwu** [comment]
@novos40 
 “Therefore when WRITTEN_AMOUNT event comes after CHANNEL_CLOSED event”

In the same tcp connection, event is serially sent or received, why WRITTEN_AMOUNT event could comes after CHANNEL_CLOSED event？

**Comment by novos40** [comment]
Only external events are sent serially. Events generated by the system itself (e.g. WRITTEN_AMOUNT) can be sent out of order. Besides this only happens under heavy load when, apparently, thread generating this event got preempted for a long enough time to be "late". I don't know if it was changed/fixed in netty 4 as, I think, there is a message queue and assigned [same] thread to process it, but it is an observable behavior in netty 3.

**Comment by normanmaurer** [comment]
Netty 3.x is end of life. So let me close this
